### ISSUE TITLE
Logging and output for extremely long running problems

### ISSUE THREAD

**Issue by janosg** [OpenSourceEconomics/estimagic#71]
## Current situation

The optimization functions return the optimal parameters and some additional information. After #54 this will be just one object. Moreover, after #45 we will have a persistent log of parameters.

## Desired situation

When optimization takes very long (between a few days and potentially weeks) and we run different optimizations in parallel, it is very easy to lose track. It has definitely happened to me that I found an optimization result on my computer and had difficulties to find out to which of many models it belonged. 

In the worst case the script that started the optimization does not exist anymore or we forgot the commit we would have to check out to get it. 

Therefore, the result of the optimization functions should (in addition to what we have currently) contain the following information:

- The start and end time of the optimization
- All arguments to `maximize` or `minimize`
- All `criterion_args` and `criterion_kwargs`
- The conda environment (exported via condas python api)
- The complete parameter log
- Optionally: additional information the user provided and that help him to reproduce the same result (e.g. the commit at which he started the optimization or a message that explains why this particular optimization was started.)

## Implementation

Collecting the data is easy. The main question is in what format we want to store it on disk.

Ideally everything would be in one "immutable" file and not just a directory where a user could save or delete files manually.  `.pickle` is not a good idea because some pickle objects will only load in the right environment and the information on the environment is inside the pickle file. `.json` won't work because we have some binary objects. My current favorite is something like a zipped folder but I'm open to other ideas. Inside this folder we could choose the best file type for each information. The environment would be a `.yml` file, for all dictionaries and lists we could first try to store them as `.yml` and only if that fails resort to `.pickle`. 

This is based on a conversation with @fediskhakov. Feel free to comment below. Other people can also add ideas and wishes. Right now there are more pressing issues in estimagic but this will be a nice addition in the long run.

**Comment by hmgaudecker** [comment]
Very nice. But... One place only: the database seems to be the right place for this. You can throw in any type of object there. 

I would put throw the current commit and the output of git diff in there, too. Maybe filtering out too large diffs.

If you are obsessed about immutability, you may be able to achieve that (but probably need a more complex backend than sqlite), but I would trust the user there.

**Comment by janosg** [comment]
I always thought of a database as some kind of rectangular data format... Of course, then the database is the place. That also solves all other implementation problems I could think of. Maybe @raholler wants to give a very brief hackathon presentation on databases next month?

Adding the current commit and the git diff automatically is a really cool idea! 

I don't think we have to be too extreme about the immutability. What I want is an object that is somehow atomic and not mutable by accident.

**Comment by raholler** [comment]
> I always thought of a database as some kind of rectangular data format... Of course, then the database is the place. That also solves all other implementation problems I could think of. Maybe @raholler wants to give a very brief hackathon presentation on databases next month?


Next month, will be tough for me, too much to do, except you want me to present it before implementing it in estimagic. Then you may have to wait a little longer for the implementation though. Also Christian already asked me to present something on the digitization stuff I have been doing, but I guess we could move this definetely to november.

**Comment by janosg** [comment]
Sure, November is also ok and we should wait until you finish the implementation in estimagic.

**Comment by tobiasraabe** [comment]
Tackled in #102.

**Comment by tobiasraabe** [comment]
Also tackled in #102?

**Comment by janosg** [comment]
Not completely. We do not log a git diff and conda environment yet. Let's leave it open.

**Comment by janosg** [comment]
We have come a long way. The rest is not a priority because it would only be needed for projects with a very bad workflow.
